File: sort/merge_sort/R/mergeSort.R
 merge <- function(x,y){
    if (is.na(x[1])) return(y)
    else if (is.na(y[1])) return(x)
    else if (x[1]<y[1]) return(c(x[1],merge(x[-1],y)))
    else return(c(y[1],merge(x,y[-1])))
 }

 division <- function(x){
    if (is.na(x[3])) return(cbind(x[1],x[2]))
    else
 return(cbind(c(x[1],division(x[-c(1,2)])[,1]),c(x[2],division(x[-c(1,2)])[,2]))) 

 }

 mergesort <- function(x){
    if (is.na(x[2])) return(x)
    else{
      print(x)
      t=division(x)
      return(merge(mergesort(t[,1]),mergesort(t[,2])))
    }
 }



 mergesort <- function(x) {
   n <- length(x)
   if (n < 2) return(x)

   # split x into two pieces of approximately equal size, x1 and x2

   x1 <- x[1:(n %/% 2)]
   x2 <- x[(n %/% 2 + 1):n]

   # sort each of the pieces
   x1 <- mergesort(x1)
   x2 <- mergesort(x2)

   # merge them back together
   result <- c()
   i <- 0
   while (length(x1) > 0 && length(x2) > 0) {
     # compare the first values
     if (x1[1] < x2[1]) {
       result[i + 1] <- x1[1]
       x1 <- x1[-1]
     } else {
      result[i + 1] <- x2[1]
       x2 <- x2[-1]
     }
     i <- i + 1
   }

   # put the smaller one into the result
   # delete it from whichever vector it came from
   # repeat until one of x1 or x2 is empty
   # copy both vectors (one is empty!) onto the end of the results
   result <- c(result, x1, x2)
   result
 }

File: sort/merge_sort/python/mergesort.py


# two sorted lists to merge
def merge(a: list, b: list) -> list:
    """

    :param a: left list to be merged
    :param b: right list to be merged
    :return: merged, sorted list
    """
    result = []  # initialise empty list to collect result
    # while both lists are not empty, add values to result in increasing order
    while len(a) > 0 and len(b) > 0:
        if a[0] <= b[0]:
            result.append(a[0])
            del a[0]
        else:
            result.append(b[0])
            del b[0]
    # either left or right may have elements left - consume them.
    while len(a) > 0:
        result.append(a[0])
        del a[0]
    while len(b) > 0:
        result.append(b[0])
        del b[0]
    # return sorted list
    return result


# conducts Merge Sort on provided list in increasing order
def merge_sort(l: list) -> list:
    """
    Sorts numeric or alphabetic list using Merge Sort

    :param l: list of values / strings to sort
    :return: list sorted in increasing order
    """
    if len(l) <= 1:
        return l

    # split the list into two approximately equally-sized lists (len max differ by 1)
    mid = len(l) // 2
    a = merge_sort(l[:mid])
    b = merge_sort(l[mid:])
    # the merge() function merges the left and right sublists
    return merge(a, b)


def main():
    # test numeric merge sort here
    my_list = [35, 12, 78, 100, 2, 39, 55, 9, 88, 77, 1000, 1, 23, 500, 46]
    my_list_sorted = merge_sort(my_list)
    print(my_list_sorted)

    # test alphabetic merge sort here
    my_alist = ['zebra', 'laptop', 'canon', 'explode', 'apple', 'dear', 'thomas', 'herbs', 'kelp', 'frog']
    my_alist_sorted = merge_sort(my_alist)
    print(my_alist_sorted)


if __name__ == "__main__":
    main()
